src/cgroups/v2/systemd_manager.rs

```diff
@@ -0,0 +1,304 @@
+                anyhow!("Invalid slice name: {}", slice);
+            }
+            // Append the component to the path and to the prefix.
+            path = format!("{}/{}{}{}", path, prefix, component, suffix);
+            prefix = format!("{}{}-", prefix, component);
+        }
+        Ok(Path::new(&path).to_path_buf())
+    }
+
+    // get_cgroups_path generates a cgroups path from the one provided by the user via cgroupsPath.
+    // an example of the final path: "/machine.slice/docker-foo.scope"
+    fn get_cgroups_path(&self) -> Result<PathBuf> {
+        // the root slice is under 'machine.slice'.
+        let mut slice = Path::new("/machine.slice").to_path_buf();
+        // if the user provided a '.slice' (as in a branch of a tree)
+        // we need to "unpack it".
+        if !self.cgroups_path.parent.is_empty() {
+            slice = self.expand_slice(self.cgroups_path.parent.clone())?;
+        }
+        let unit_name = self.get_unit_name();
+        let cgroups_path = slice.join(unit_name);
+        Ok(cgroups_path)
+    }
+
+    /// create_unified_cgroup verifies sure that *each level* in the downward path from the root cgroup
+    /// down to the cgroup_path provided by the user is a valid cgroup hierarchy,
+    /// containing the attached controllers and that it contains the container pid.
+    fn create_unified_cgroup(&self, pid: Pid) -> Result<PathBuf> {
+        let cgroups_path = self.get_cgroups_path()?;
+        let full_path = self.root_path.join_absolute_path(&cgroups_path)?;
+        let controllers: Vec<String> = self
+            .get_available_controllers(common::DEFAULT_CGROUP_ROOT)?
+            .into_iter()
+            .map(|c| format!("{}{}", "+", c.to_string()))
+            .collect();
+
+        // Write the controllers to the root_path.
+        Self::write_controllers(&self.root_path, &controllers)?;
+
+        let mut current_path = self.root_path.clone();
+        let mut components = cgroups_path.components().skip(1).peekable();
+        // Verify that *each level* in the downward path from the root cgroup
+        // down to the cgroup_path provided by the user is a valid cgroup hierarchy.
+        // containing the attached controllers.
+        while let Some(component) = components.next() {
+            current_path = current_path.join(component);
+            if !current_path.exists() {
+                fs::create_dir(&current_path)?;
+                fs::metadata(&current_path)?.permissions().set_mode(0o755);
+            }
+
+            // last component cannot have subtree_control enabled due to internal process constraint
+            // if this were set, writing to the cgroups.procs file will fail with Erno 16 (device or resource busy)
+            if components.peek().is_some() {
+                Self::write_controllers(&current_path, &controllers)?;
+            }
+        }
+
+        write_cgroup_file(full_path.join(CGROUP_PROCS), &pid.to_string())?;
```

Comment:
You can leave off to_string here...write_cgroup_file takes all types that implement ToString and will call it for you. Currently this means to_string is called twice. Once by you and once by write_cgroup_file.

Reply:
Nit: prefix write_cgroup_file with common to be consistent with the other usages

Reply:
Done.